runtime/libffi/z/sysvz64.s

```diff
@@ -432,51 +480,360 @@ DARGF    DS 0H                l_DOUBLE in arg area
+         LH 15,10(,15)          type->elements[1]->type
+         CFI 15,1               is it a double?
+         BE STFPCHKD2
+         B STRCT2
+
+
+*check if there are more elements
+*if not we have a special case
+*I think this is to handle complex types
+*in this case we pass the struct
+*in the first 2 free FPRs, just check this
+*against r14 the free FPR counter
+STFPCHKF2 DS 0H
+         LG 15,16(,11)          type->elements
+         LA 15,16(,15)          type->elements[0]
+         LG 15,0(,15)
+         CFI 15,0
+         BNE STRCT2             if no, not special case
+
+         LG  11,0(10,5)
+         LG 15,0(,11)           type->size
+
+         CFI 14,4               no FPRs are available
+         BNL STRCT2             if yes, not special case
+
+         CFI 14,3               1 FPR is available                
+         BNL STFPCHKF26         use fpr6 + memory
+
+         CFI 14,2               2 FPRs available
+         BE STFPCHKF246         use fpr4 + fpr6
+
+         CFI 14,1               3 FPRs available
+         BE STFPCHKF224         use fpr2 + fpr4
+
+         CFI 14,0               4 FPRs available
+         BE STFPCHKF202         use fpr0 + fpr2
+
+
+STFPCHKF26 DS 0H
+         LE 6,0(6,13)           first float will go in fpr
+         AFI 0,1                one less GPR available
+         AFI 14,1               one less FPR available
+
+         AFI 6,4                bump local store 4 bytes
+         AFI 15,-4              we've handled 4 bytes
+         AFI 7,4                advance 4 bytes into argarea
+
+         B STRCTLP              copy rest of struct
+
+
+STFPCHKF202 DS 0H
+         LE 0,0(6,13)           first float
+         LE 2,4(6,13)           second float
+         AFI 0,2                one less GPR available
+         AFI 14,2               two less FPRs available
+
+         AFI 6,8                bump arg area by 8 bytes
+         AFI 15,-8              we've handled 8 bytes (all the struct)
+         AFI 7,8                bump local storage by 8 bytes
+
+         B STRCTLP
+
+STFPCHKF224 DS 0H
+         LE 2,0(6,13)           first float
+         LE 4,4(6,13)           second float
+         AFI 0,2                one less GPR available
+         AFI 14,2               two less FPRs available
+
+         AFI 6,8                bump arg area by 8 bytes
+         AFI 15,-8              we've handled 8 bytes (all the struct)
+         AFI 7,8                bump local storage by 8 bytes
+
+         B STRCTLP
+
+
+STFPCHKF246 DS 0H
+         LE 4,0(6,13)           first float
+         LE 6,4(6,13)           second float
+         AFI 0,2                one less GPR available
+         AFI 14,2               two less FPRs available
+
+         AFI 6,8                bump arg area by 8 bytes
+         AFI 15,-8              we've handled 8 bytes (all the struct)
+         AFI 7,8                bump local storage by 8 bytes
+
+         B STRCTLP
+
+
+
+*check if there are more elements
```

Comment:
May be we should consolidate this comment with the previous one explaining why complex float types needed special handling.